File: example/reading/python/template.py
#!/usr/bin/env python3

import pandas as pd
import numpy as np
import statsmodels.formula.api as smf

if __name__ == '__main__':
    # read data
    df = pd.read_csv('./data.csv')

    # take the first N participants to simulate stopping condition
    df = df[:{{sample_size}}]

    # calculate reading speed in WPM
    df['speed'] = df.apply(lambda row: row.num_words / row['{{rt}}'] * 60000,
                           axis=1)

    # convert education level into an ordinal variable
    edu_order = ['pre-high school', 'high school', 'professional school',
                 'college', 'graduate school', 'PhD', 'postdoctoral']
    tp = pd.CategoricalDtype(categories=edu_order, ordered=True)
    df['edu_level'] = df.education.astype(tp).cat.codes

    # remove retake participants
    df = df[df.retake != 1]

    # remove smart phone users
    df = df[~df.device.isin({{bad_device}})]

    # remove outliers based on reading speed
    # --- (B1)
    # remove reading speed outside median + 3 x iqr
    iqr = np.subtract(*np.percentile(df.speed, [75, 25]))
    cutoff_high = np.median(df.speed) + 3 * iqr

    # --- (B2)
    # remove reading speed outside mean + 2 x std
    cutoff_high = np.mean(df.speed) + 2 * np.std(df.speed)

    # --- (C)
    cutoff_low = {{min_wpm}}
    df = df[df.speed <= cutoff_high]
    df = df[df.speed >= cutoff_low]

    # drop NA rows
    df = df.dropna()

    # log-normalize speed
    df['log_speed'] = np.log(df.speed)

    # decision: whether to bin dyslexia or not
    df.dyslexia = df['{{dyslexia}}']

    # make dyslexia a categorical variable
    df.dyslexia = df.dyslexia.astype('category')

    # remove trials based on comprehension < 2/3
    # --- (D1)
    # just remove trials
    df = df[df.correct_rate > 0.6]

    # --- (D2)
    # drop entire participants
    bad_uuid = set()
    for i, row in df.iterrows():
        if row.correct_rate < 0.6:
            bad_uuid.add(str(row.uuid))
    df = df[~df.uuid.isin(bad_uuid)]

    # --- (F1)
    # fit a linear mixed effects model
    fml = '{{formula}}'
    model = smf.mixedlm(fml, df, groups=df.uuid).fit()
    print(model.summary())

    # --- (F2)
    # fit a multinomial logit model to accuracy
    df['acc'] = 3 - pd.Categorical(df.correct_rate).codes
    fml = 'acc ~ page_condition*dyslexia_bin'
    model = smf.mnlogit(fml, df).fit()
    print(model.summary())

File: example/reading/r/template.R
#!/usr/bin/env Rscript
# --- (BOBA_CONFIG)
{
  "graph": [
    "RC->LM1->O1",
    "RC->LM2->O2",
    "OLR1->O1",
    "OLR2->O2"
  ],
  "decisions": [
    {"var": "brmsfamily", "options": ["shifted_lognormal", "lognormal"]}
  ],
  "outputs": [
    {"name": "aic/waic", "value": "aic"}
  ],
  "before_execute": "cp ../../data.csv ./code/ && mkdir results"
}
# --- (END)

library(readr)
library(lmerTest)
library(car)
library(psych)
library(scales)
library(brms)
library(ordinal)

speed_data <- read_csv('data.csv')

# calculate reading speed in WPM
speed_data$speed <- speed_data$num_words/(speed_data$adjust_rt/60000)

# remove retake participants
speed_data <- subset(speed_data, retake != 1)

# remove outliers
iqr = IQR(speed_data[speed_data$dyslexia_bin == 0,]$speed,na.rm=TRUE)
cutoff_high = median(speed_data$speed) +3*iqr #3*iqr=645, cutoff_high = 928

# remove trials based on speed
result_analysis <- speed_data[! speed_data$speed > cutoff_high, ]
result_analysis <- result_analysis[ ! result_analysis$speed < 10,]

# remove smartphone users
# removed 64 smartphone users, 363 trials
result_analysis <- result_analysis[! result_analysis$device == 'smartphone',]

# wrangle variables
result_analysis$log_speed <- log(result_analysis$speed)
result_analysis$dyslexia = as.factor(result_analysis$dyslexia)
result_analysis$correct_num = round(result_analysis$correct_rate * 3, 0)
result_analysis$acc = result_analysis$correct_num + 1
result_analysis$correct_num = as.factor(result_analysis$correct_num)

# --- (RC)
# remove trials based on comprehension < 2/3
# removed 111 trials
result_analysis <- result_analysis[ ! result_analysis$correct_rate < .6,]

# --- (LM1)
# fit linear mixed model
model <- lmer(log_speed ~ page_condition*dyslexia + img_width + num_words + age + english_native + (1 | uuid),
              data = result_analysis)
print.odds = FALSE

# --- (OLR1)
# fit ordinal logistic regression using accuracy as DV
model <- clmm(correct_num ~ page_condition*dyslexia + num_words + age + english_native + (1 | uuid),
              data=result_analysis)
print.odds = TRUE

# --- (LM2)
# fit bayesian model
model <- brm(speed ~ page_condition*dyslexia + img_width + num_words + age + english_native + (1 | uuid),
             data = result_analysis, family = {{brmsfamily}}(), file = '../results/brmsfit_{{_n}}',
             save_all_pars = TRUE, silent = TRUE, refresh = 0, seed = 0,
             chains = 4, cores = 4, iter = 1000)

# --- (OLR2)
# fit bayesian model to accuracy
model <- brm(acc ~ page_condition*dyslexia + num_words + age + english_native + (1 | uuid),
             data = result_analysis, family = cumulative(), file = '../results/brmsfit_{{_n}}',
             save_all_pars = TRUE, silent = TRUE, refresh = 0, seed = 0,
             chains = 4, cores = 4, iter = 1000)

# --- (O1)
aic = AIC(model)
sink('../results/summary_{{_n}}.txt')
summary(model)

if(print.odds){
    print("Odds ratio:")
    exp(coef(model))
}

# --- (O2)
# evaluate fit
aic = waic(model)$waic

# output resultsf
sink('../results/summary_{{_n}}.txt')
summary(model)
sink()
pdf(file="../results/plots_{{_n}}.pdf")
plot(model)
marginal_effects(model)
